astrobee/behaviors/inspection/src/camera_projection.cc

```diff
@@ -0,0 +1,444 @@
+                                    << p_far(2, 0) / p_far(3, 0));
+        ROS_DEBUG_STREAM("p_far 2 " << p_far(0, 1) / p_far(3, 1) << " " << p_far(1, 1) / p_far(3, 1) << " "
+                                    << p_far(2, 1) / p_far(3, 1));
+        ROS_DEBUG_STREAM("p_far 3 " << p_far(0, 2) / p_far(3, 2) << " " << p_far(1, 2) / p_far(3, 2) << " "
+                                    << p_far(2, 2) / p_far(3, 2));
+        ROS_DEBUG_STREAM("p_far 4 " << p_far(0, 3) / p_far(3, 3) << " " << p_far(1, 3) / p_far(3, 3) << " "
+                                    << p_far(2, 3) / p_far(3, 3));
+      }
+      // Eliminate point
+      return false;
+    }
+
+
+    return true;
+  }
+
+
+  bool CameraView::GetCamXYFromPoint(const geometry_msgs::Point point, int& x, int& y) {
+    // Initialize x,y
+    x = 0; y = 0;
+    // Get current camera position if it's not given
+    geometry_msgs::TransformStamped robot_pose;
+    try {
+      robot_pose = tf_buffer_.lookupTransform("world", "body", ros::Time(0), ros::Duration(1.0));
+    } catch (tf2::TransformException &ex) {
+      ROS_ERROR("Failed getting transform: %s", ex.what());
+      return false;
+    }
+    return GetCamXYFromPoint(msg_conversions::ros_transform_to_ros_pose(robot_pose.transform), point, x, y);
+  }
+
+  // Get point from camera pixel location and point cloud
+  bool CameraView::GetPointFromXYD(const sensor_msgs::PointCloud2 pCloud, const int x, const int y,
+                                   geometry_msgs::Point& point) {
+    // Convert from u (column / width), v (row/height) to position in array
+    // where X,Y,Z data starts
+    int arrayPosition = x*pCloud.row_step + y*pCloud.point_step;
+
+    // compute position in array where x,y,z data start
+    int arrayPosX = arrayPosition + pCloud.fields[0].offset;  // X has an offset of 0
+    int arrayPosY = arrayPosition + pCloud.fields[1].offset;  // Y has an offset of 4
+    int arrayPosZ = arrayPosition + pCloud.fields[2].offset;  // Z has an offset of 8
+
+    float X = 0.0;
+    float Y = 0.0;
+    float Z = 0.0;
+
+    memcpy(&X, &pCloud.data[arrayPosX], sizeof(float));
+    memcpy(&Y, &pCloud.data[arrayPosY], sizeof(float));
+    memcpy(&Z, &pCloud.data[arrayPosZ], sizeof(float));
+
+    // put data into the point p
+    point.x = X;
+    point.y = Y;
+    point.z = Z;
+
+    return true;
+  }
+  bool CameraView::InsideTarget(std::vector<int> vert_x, std::vector<int> vert_y, int test_x, int test_y) {
+    int i, j;
+    bool c = false;
+    int n_vert = 4;
+    for (i = 0, j = n_vert-1; i < n_vert; j = i++) {
+      if (((vert_y[i] > test_y) != (vert_y[j] > test_y)) &&
+       (test_x < (vert_x[j] - vert_x[i]) * (test_y - vert_y[i]) / (vert_y[j]-vert_y[i]) + vert_x[i]) )
+         c = !c;
+    }
+    return c;
+  }
+
+  double CameraView::GetDistanceFromTarget(const geometry_msgs::Pose point, std::string depth_cam_name, double size_x,
```

Comment:
It doesn't take only the corners, it takes in all the pixels that are inside the target area (that's what the "InsideTarget" function classifies - if defines if a point is inside or outside a polygon defined by the corners).

It also takes into account if any corners of the target are outside the image as the for-loop goes through all the pixels in the image, but passes onto the "InsideTarget" corners that might be out-of-image bounds. (also note that the function does return false if the point is out-of-bounds, but in the case we don't mind that).

As for not taking the actual distance and only the z-coord, that's really interesting, I can change that and rename the function to make clear of the output.

Reply:
I see, I misread part of the lower section.

However, the point about the error logic remains. The `GetCamXYFromPoint()` method can return `false` for many reasons. If the body-to-camera transform lookup fails, it returns (x, y) = (0, 0), which is nonsense for `InsideTarget` checking. Another case when it gives `false` is when the point is behind the camera, in which case the (x, y) coordinates can again be nonsensical.

And it was a genuine question why we need to estimate the average depth over the entire target. Is there much reason to think that will be more robust than the simpler approach of using a single depth value from the center of the target?

Reply:
I've addressed this in a comment on the code. At this point, the target is validated to be within the inspection camera (sci_cam). So even if it returns false here is is due to a smaller fov. In case the camera query fails (which is very unlikely), then an error will be printed on the cmd line for the isaac operator to see and act upon it (I could have put it to be FATAL, but if this ever happens online, it's better to have the chance to resend the command than having to restart the node by default).

Reasons to estimate average: irregular target to inspect, taking an average will make sure that the camera is focused the best it can on the entire object; reduces noise from haz_cam measurements. In general with things that imply uncertainty I'm a proponent of taking more than one measurement.